### PR TITLE
Add support for @IntoMap

### PR DESCRIPTION
**Pull request by JakeWharton**
Need a test for `unwrapValue=false` but that can come later.

### CODE REVIEW

File: reflect/src/main/java/dagger/reflect/BindingMap.java [hunk 1]
```diff
@@ -98,6 +99,23 @@ Builder addIntoSet(Key key, Binding binding) {
       return this;
     }
 
+    Builder addIntoMap(Key key, Object mapKey, Binding mapValueBinding) {
```
Comment (JakeWharton) [inline reflect/src/main/java/dagger/reflect/BindingMap.java]: Maybe using 'entry' would be a good prefix. Similar to 'element' for sets.

Comment (burntcookie90) [reply reflect/src/main/java/dagger/reflect/BindingMap.java]: `bindingKey`?

File: reflect/src/main/java/dagger/reflect/ReflectiveModuleParser.java [hunk 2]
```diff
@@ -69,11 +72,44 @@ static void parse(Class<?> moduleClass, @Nullable Object instance,
...
+      keyValue = Reflection.tryInvoke(keyAnnotation, keyAnnotationMethod);
+      if (keyValue == null) {
+        throw new AssertionError(); // Not allowed by the Java language specification.
+      }
+    } else {
+      keyType = keyAnnotationType;
+      keyValue = keyAnnotation;
+    }
+
+    Key key = Key.of(valueKey.qualifier(),
```
Comment (burntcookie90) [inline reflect/src/main/java/dagger/reflect/ReflectiveModuleParser.java]: key key key key key key

File: reflect/src/main/java/dagger/reflect/UnlinkedMapBinding.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    for (Map.Entry<Object, Binding> entryBinding : entryBindings.entrySet()) {
+      LinkedBinding<Object> linkedBinding =
+          (LinkedBinding<Object>) entryBinding.getValue().link(linker);
+      linkedBindings.put(entryBinding.getKey(), linkedBinding);
+    }
+    return new LinkedMapBinding<>(linkedBindings);
+  }
+
+  @Override public String toString() {
+    return "Set" + entryBindings;
```
Comment (burntcookie90) [inline reflect/src/main/java/dagger/reflect/UnlinkedMapBinding.java]: "Set" should be "Map"

File: reflect/src/main/java/dagger/reflect/BindingMap.java (hunk 1 above)
Comment (JakeWharton) [reply reflect/src/main/java/dagger/reflect/BindingMap.java]: I could also just write some docs...

File: reflect/src/main/java/dagger/reflect/UnlinkedMapBinding.java (hunk 3 above)
Comment (JakeWharton) [reply reflect/src/main/java/dagger/reflect/UnlinkedMapBinding.java]: Applied locally

File: reflect/src/main/java/dagger/reflect/BindingMap.java (hunk 1 above)
Comment (JakeWharton) [reply reflect/src/main/java/dagger/reflect/BindingMap.java]: Settled on entryKey and entryValueBinding as least confusing... for now.

**Pull request merged**